### PR TITLE
PUBDEV-8428 backward selection

### PR DESCRIPTION
**Pull request by wendycwong**
This PR completes the task in JIRA: https://h2oai.atlassian.net/browse/PUBDEV-8428

I completed the following:
1. Java backend to support backward selection
2. Junit tests;
3. Python/R unit tests
4. Add documentation to help team do review for me.

### CODE REVIEW

**Comment by wendycwong** [comment]
Attached the pdf here:
[ModelSelectionTutorial.pdf](https://github.com/h2oai/h2o-3/files/7866772/ModelSelectionTutorial.pdf)

**Comment by neemah2o** [comment]
@wendycwong , I am probably not much help with the Java code, but I read your tutorial. If you decide to add parts of to the docs, then consider my suggestions
[ModelSelectionTutorial.pdf](https://github.com/h2oai/h2o-3/files/7874051/ModelSelectionTutorial.pdf)
:

File: h2o-algos/src/main/java/hex/modelselection/ModelSelection.java [hunk 1]
```diff
@@ -80,70 +78,121 @@ public boolean havePojo() {
...
-                model._output._best_model_predictors = new String[_parms._max_predictor_number][];
-                model._output._coefficient_names = new String[_parms._max_predictor_number][];
+                model._output._best_model_ids = backward.equals(_parms._mode) ? new Key[_numPredictors] 
+                        : new Key[_parms._max_predictor_number];
+                if (backward.equals(_parms._mode)) {
+                    model._output._coef_p_values = new double[_numPredictors][];
+                    model._output._z_values = new double[_numPredictors][];
+                } else {
+                    model._output._best_r2_values = new double[_parms._max_predictor_number];
+                }
```
Comment (maurever) [inline h2o-algos/src/main/java/hex/modelselection/ModelSelection.java:177]:
```suggestion
                if (backward.equals(_parms._mode)) {
                    model._output._best_model_ids = new Key[_numPredictors] 
                    model._output._coef_p_values = new double[_numPredictors][];
                    model._output._z_values = new double[_numPredictors][];
                } else {
                    model._output._best_model_ids = new Key[_parms._max_predictor_number];
                    model._output._best_r2_values = new double[_parms._max_predictor_number];
                }
```

File: h2o-algos/src/main/java/hex/modelselection/ModelSelection.java [hunk 2]
```diff
@@ -80,70 +78,121 @@ public boolean havePojo() {
...
-            error("_family", "ModelSelection only supports Gaussian family");
+        _predictorNames = extractPredictorNames(_parms, _dinfo, _foldColumn);
+        _numPredictors = _predictorNames.length;
+
+        if (maxr.equals(_parms._mode) || allsubsets.equals(_parms._mode)) { // check for maxr and allsubsets
+            if (nclasses() > 1)
+                error("response", "'allsubsets' and 'maxr' only works with regression.");
+            
+            if (!(AUTO.equals(_parms._family) || gaussian.equals(_parms._family)))
+                error("_family", "ModelSelection only supports Gaussian family");
```
Comment (maurever) [inline h2o-algos/src/main/java/hex/modelselection/ModelSelection.java:124]:
```suggestion
                error("_family", "ModelSelection only supports Gaussian family for 'allsubset' and 'maxr' mode.");
```

File: h2o-algos/src/main/java/hex/schemas/ModelSelectionModelV3.java [hunk 3]
```diff
@@ -14,8 +14,14 @@
         @API(help="Best predictor subset names for each subset size.")
         String[][] best_model_predictors; // store for each predictor number, the best model predictors
         
-        @API(help="R2 values of all possible predictor subsets.")
+        @API(help="R2 values of all possible predictor subsets.  Only for model='allsubsets' or 'maxr'")
```
Comment (maurever) [inline h2o-algos/src/main/java/hex/schemas/ModelSelectionModelV3.java:17]:
```suggestion
        @API(help="R2 values of all possible predictor subsets.  Only for model='allsubsets' or 'maxr'.")
```

File: h2o-algos/src/main/java/hex/schemas/ModelSelectionModelV3.java [hunk 4]
```diff
@@ -14,8 +14,14 @@
...
         double[] best_r2_values;  // store the best R2 values of the best models with fix number of predictors
+
+        @API(help="p-values of chosen predictor subsets at each subset size. Only for model='backward'")
```
Comment (maurever) [inline h2o-algos/src/main/java/hex/schemas/ModelSelectionModelV3.java:20]:
```suggestion
        @API(help="p-values of chosen predictor subsets at each subset size. Only for model='backward'.")
```

File: h2o-algos/src/main/java/hex/schemas/ModelSelectionModelV3.java [hunk 5]
```diff
@@ -14,8 +14,14 @@
...
+        double[][] coef_p_values;
+
+        @API(help="z-values of chosen predictor subsets at each subset size. Only for model='backward'")
```
Comment (maurever) [inline h2o-algos/src/main/java/hex/schemas/ModelSelectionModelV3.java:23]:
```suggestion
        @API(help="z-values of chosen predictor subsets at each subset size. Only for model='backward'.")
```

File: h2o-algos/src/test/java/hex/modelselection/BackwardSelectionTests.java [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+                    String frameValue = resultF.vec(frameNames[cInd]).stringAt(rInd);
+                    assertTrue(cellValue.equals(frameValue));
+                }
+        } finally {
+            Scope.exit();
+        }
+    }
+    
+    // test with using p-values to stop model building
+    // compare with manually built frames, check coefficients, z-values, p-values
```
Comment (maurever) [inline h2o-algos/src/test/java/hex/modelselection/BackwardSelectionTests.java:59]:
```suggestion
    /** 
      * test with using p-values to stop model building
      * compare with manually built frames, check coefficients, z-values, p-values
      */
```

File: h2o-py/h2o/estimators/model_selection.py [hunk 7]
```diff
@@ -271,13 +295,22 @@ def __init__(self,
                capability
                Defaults to ``0``.
         :type nparallelism: int
-        :param max_predictor_number: Maximum number of predictors to be considered when building GLM models.  Defaiult
-               to 1.
+        :param max_predictor_number: Maximum number of predictors to be considered when building GLM models.  Default to
```
Comment (maurever) [inline h2o-py/h2o/estimators/model_selection.py:298]:
```suggestion
        :param max_predictor_number: Maximum number of predictors to be considered when building GLM models.  Defaults to
```

File: h2o-py/h2o/estimators/model_selection.py [hunk 8]
```diff
@@ -1033,20 +1145,51 @@ def max_predictor_number(self, max_predictor_number):
...
+    def min_predictor_number(self, min_predictor_number):
+        assert_is_type(min_predictor_number, None, int)
+        self._parms["min_predictor_number"] = min_predictor_number
+
     @property
     def mode(self):
         """
-        Mode: used to choose model selection algorithm to use,
+        Mode:  Used to choose model selection algorithms to use.  Options include 'allsubsets' for all subsets'maxr' for
+        MaxR'backward' for backward selection
```
Comment (maurever) [inline h2o-py/h2o/estimators/model_selection.py:1167]:
```suggestion
        Mode:  Used to choose model selection algorithms to use.  Options include 'allsubsets' for all subsets, 'maxr' for
        MaxR, 'backward' for backward selection.
```

File: h2o-r/h2o-package/R/modelselection.R [hunk 9]
```diff
@@ -103,9 +111,13 @@
...
 #' @param nparallelism number of models to build in parallel.  Default to 0.0 which is adaptive to the system capability Defaults to
 #'        0.
-#' @param max_predictor_number Maximum number of predictors to be considered when building GLM models.  Defaiult to 1. Defaults to 1.
-#' @param mode Mode: used to choose model selection algorithm to use,  Must be one of: "allsubsets", "maxr". Defaults to
-#'        maxr.
+#' @param max_predictor_number Maximum number of predictors to be considered when building GLM models.  Default to 1. Defaults to 1.
+#' @param min_predictor_number For mode = 'backward' only.  Minimum number of predictors to be considered when building GLM models starting
+#'        with all predictors to be included.  Default to 1. Defaults to 1.
+#' @param mode Mode:  Used to choose model selection algorithms to use.  Options include 'allsubsets' for all subsets'maxr'
+#'        for MaxR'backward' for backward selection Must be one of: "allsubsets", "maxr", "backward". Defaults to maxr.
```
Comment (maurever) [inline h2o-r/h2o-package/R/modelselection.R:118]:
```suggestion
#' @param mode Mode:  Used to choose model selection algorithms to use.  Options include 'allsubsets' for all subsets, 'maxr'
#'        for MaxR, 'backward' for backward selection Must be one of: "allsubsets", "maxr", "backward". Defaults to "maxr".
```

Note: Need to be fixed in gen_modelselection.py I guess.

**Review by maurever — APPROVED** [review]
LGTM, thanks @wendycwong. Nice test coverage. I have just a few suggestion.

**Pull request merged**